dashboard/app/models/user.rb

```diff
@@ -2261,6 +2261,41 @@ def providers
     end
   end
 
+  # Returns number of times a user has attempted to join a section in the last 24 hours
+  # Returns nil if no section join attempts
+  def num_section_attempts
+    properties['section_attempts'] || 0
+  end
+
+  # There aare two possible states in which we would want to reset section attempts
+  # 1) Initialize for the first time 2) 24 hours have passed since last attempt
+  def reset_section_attempts?
+    !num_section_attempts || !properties['section_attempts_last_reset'] || num_section_attempts == 0 || (DateTime.now - DateTime.parse(properties['section_attempts_last_reset'])).to_i > 0
+  end
+
+  def reset_section_attempts
+    properties['section_attempts'] = 0
+    properties['section_attempts_last_reset'] = DateTime.now.to_s
+    save
+  end
+
+  def reset_section_attempts_if_needed
+    # Failed section attempts reset after 24 hours
+    reset_section_attempts if reset_section_attempts?
+  end
+
+  def display_captcha?
```

Comment:
I'm not really sure how to fix this, but display_captcha? has a potentially unexpected side effect of saving the user object (which in turn triggers other code).  At the very least, this should be mentioned in the header comment, but perhaps there's a way to structure the code to make this function more side-effect-free?

Reply:
I've really struggled with this question myself. It's a very good question, and I would love to maybe brainstorm together on it. The issue I have is a user could join a section for the first time and when we check whether or not to display the captcha, all of the properties related to joining a section would return `nil`. We could probably account for that case, but then there's the question of a user joining a section for the 4th time in 25 hours. In the 25th hour, we'll need to reset the values before determining whether or not to display the captcha.

Maybe `increment_section_attempts` is the only place we call `save` and display_captcha can reset section attempts where necessary and return the resulting boolean without actually changing values in the db. Then, when we actually account for the fact that they pressed "Submit" on joining a section, we can handle any reseting. That was a lot... let me know if it made sense or if you would like to discuss it further.

Reply:
Make it a pure function - want it to be clear that it has no side effects.
Try not to alter state in a method returning a boolean:
1) if reset was over a day ago, return false; then, 
2) if number of attempts is greater than or equal to three, return true; else, return false.

Then, expand code in the increment method and note that it should only be called on a user object that has been persisted –  can add this check explicitly.